### DISCUSSION TITLE
Some UX feedback

### DISCUSSION THREAD

**Discussion by jcklpe** [mozilla/hubs#3153] [General Questions & Discussion] [answered]
# Preface and my background

I'm a professional UX designer, and have been involved in the local Austin Texas VR scene since the DK2 and [am experienced with 3D art creation and several asset creation pipelines](https://sketchfab.com/thedonquixotic). I pivoted away from the vr and video games industry a couple of years ago but still maintain an interest in it as a part of my general interest in my field and as an artist. 

I had a birthday recently. more than 5 people in a video call gets confusing really fast. So I thought: why not try Hubs? It was a great experience!

I also tried it out for a meeting of my tech-art collective (https://vurv.co) 

I'd like to get more involved if I can. I am not a particularly good developer. That's why my job title is "UX Designer". I know frontend but I'm not the kinda guy who can sit down and bust through leetcode stuff. I do have a strong interest in open source though and would like to be helpful. I'm not coming in trying to be an "idea guy" or something. Just have some feedback that might be useful, and I'm more well acquainted with the with the space and it's technical definitions than the average person or even average UX designer so I hope it might be useful. 

# How I experienced Hubs
I used hubs in the browser. Everyone else did too. The party lasted about 5 hours and had people dropping in and out. I created a room in Hubs and then shared the link to the facebook event 24 hours before the event was scheduled to start. 

# Feedback

### Proximity chat stuff
a big attraction to me about Hubs was that it had proximity voice. This attractive because it could mean that people could break off into little "discussion groups". But unlike separating out a big zoom call into mini chat rooms which are discrete and require an explicit entrance and exit, Hubs allows people to sort of wander from one group to the next in a very liquid way which is more like how a real life party works. 

This is cool but the fall off is not so great. It takes A LONG LONG TIME for a person's voice to fall off to a level where it ceases to interfere with hearing. As a result rather than having people break off into groups, it mostly mean people all stayed in one big group. 

Ways to possibly fix this would be to make there some explicit cut off point. I haven't looked at the code but I imagine that the distance of the proximity chat is some kind of continuous distance function. Perhaps there are some discrete cut off points already or perhaps there is is a bezier curve used to author the falloff a bit, but I think there should be somewhere on the lower end that cuts it off much more completely or at least that the falloff of sound is sharper at some point. 

It might be nice to make different scenes have different acoustics! Big open spaces might need more distance while tiny spaces don't. If I'm nestled in some little grotto bar I really only need to hear the people right next to me, not two tables over. 

This might in fact be something that people should be able to customize. Similar to how FPS games allow for users to dial in the sensitivity of their mouse look, it would make sense I think to allow users to dial in their hearing. This wouldn't have to confuse the user's screen, good defaults should still be the priority but IMO this is something I'd want more direct control of as a user. 

It would also be cool if there was some ability to designate "rooms" so at in creating a space on Spoke people couldn't just simply hear other people through the walls. If someone is on the second floor above me I shouldn't be able to hear them as easily as the guy 4 feet away in the same room. Running ray traced sound in the browser probably isn't super easy so might be worth considering building this into the actual level design or something, sort of like baking lighting into a scene. 

### Object controls: rotate

There really needs to be a way to rotate stuff. If there is then I couldn't figure it out. The scrollwheel thing works pretty well for size and placement of objects but if my webcam video gets donked over to the side then there's no easy way to reposition it. I can rotate my view of it, but I can't actually rotate the object for other people, which I think is very annoying. 

### Object controls: delete

I want to be able to delete stuff by right clicking on them, or by grabbing them and hitting the "delete" key. You won't believe how many times I found myself doing that. It seems very intuitive. Deleting objects by having to cycle through the list is not intuitive. It also is unclear which identical low poly cheeseburger model is which. Is it one of the 30 tiny ones that my friend spawned around the castle, or is the giant 100 foot one that he spawned into the room we are standing in and which is making it hard for everyone to see? Hmmm... sure which I could know....

### View controls

The view controls are not a good solution to the problem they're trying to solve. I get the idea of wanting view and the ability to manipulate objects in space being separate but I think having some kind of toggle key would be much better for that, with the default toggle being the head tracking and the mouse being together like a normal FPS. 

Perhaps doing this was deemed "too much for non gamers" or something of that sort. But my experience was in fact the opposite. The people who had the most difficulty were in fact the people who did not have much video game experience. The near universal wasd mouse control scheme is near universal for a reason. It works. It's basically a fundamental constraint of the medium. Cliff Bleszinski has talked about this subject in some interviews if you can dig those up. I'm not saying wasd mouse is the only way to do FP POV but it's pretty hard to do something else well. 

### Webcam head
The inclusion of the webcam is AWESOME! But it's missing just one little feature....
I want to be able to look at the person I'm talking to, and placing the webcams feeds in a way in which people can see you is a little hard. I think it would be cool if you could affix the webcam to your avatar's head. That could even be part of the visual aesthetic of the avatar like this:

![image](https://64.media.tumblr.com/09c2b7e77ee25882938b7a672240402c/0b1006aaa197101b-1c/s640x960/6ce6df90758ed7daa84f1e96f2a20873e581b46c.gif)

### Emojis

The Emojis are cute but we never used them and their taking up the space bar is too generous to the level of actual UI realestate they take up. 

Also I think having to hit space and then grab them is sort of annoying. I'd much prefer they were hotkeyed to the numbers or something like that and then simply teleported a set distance in front of me, or perhaps simply popped up over my head. That's the point of an emoji after all right? It's to tell OTHER PEOPLE your feelings, not see them yourself. If you have a huddle of 4 people and someone pops out an emoji and everyone is close together it's kinda hard to see whose it was. This seems to me to be a classic case of designing a thing in the hypothetical and not playing out the logical consequences of it with user testing beyond getting whatever positive reactions people give to the novelty of it. If this is a serious feature it needs to be more robust. Otherwise it's taking up too much space in the UX. It's cruft. 

### Hubs links
Are genius. Having persistent rooms that I can simply drop a link to people is amazing. It makes it feel like a "place" that people can drop by. It's awesome, I love it. Easily one of the best features of Hubs. 

### Jump
I'm sure ya'll have heard this before but jumping would be nice. Not having jumping feels weird. 

### Onboarding 
Onboarding to joining a new room could use a second or 3rd pass. Currently it's a little janky. It isn't clear you're in a lobby at first and I think the visual treatment could use some work.

**Comment by camelman90** [comment]
For the webcam I've experimented a little bit with that. The webcam head solution introduces confusion as people can't monitor/see their own video, unless you place it directly in front of them (but blocking their view) or create another video somewhere in the HUD.  It can also be non-apparent to the people sharing camera that it is actually on their head due to first-person perspective.

**Reply by jcklpe** [reply]
Yeah I figured that might be the reason though I think having the video in the HUD would be a good solution. In my experience, I do like being able to see what my camera is, but having to position myself so that I can see both my output and my friends output at the same time gets a little awkward. Instead of standing in a huddle talking to each other we end up sitting next to each other with our screens out in front of us: essentially replicating the zoom call UI but in 3D space and more annoying to manage. 

I think a little video feed in the corner of the HUD would be great!

**Comment by camelman90** [comment]
You can rotate objects by holding SPACE, hovering above the object and clicking the little arrows-in-a-circle icon. Hold SHIFT for even finer rotation.

You can also delete objects from this same menu.

**Reply by jcklpe** [reply]
OH! Okay there we go. Seems like it would make sense to make the controls for that fit in more with the pattern for moving in z space and size etc, but yeah that's more of a feedback on enhancement than new functionality.

**Comment by camelman90** [comment]
You can also set custom audio rolloff options when you create a scene in Spoke (hubs.mozilla.com/spoke) - the "level editor" for Mozilla Hubs. It has both linear and exponential rolloff options and a bunch more options there. It would be nice if someone could figure out a better "default" i guess, but its also dependent on the scene you used. 

You should also refer to the scene you used so someone can figure out what audio settings you based your feedback on.

**Reply by jcklpe** [reply]
I see where to change the audio rolloff for a spawned audio object in spoke but I don't see where to control the user's "hearing". 

I also probably should have divided my suggestions on audio stuff into two suggestions since they're really two different things:

1. customizable user hearing rolloff
2. customizable "audio isolation" in spoke. So the ability to define "walls" which sound can't pass through or "containers" in which sound is muffled outside of. It feels weird for interior rooms to "sound" the same as exterior rooms.

**Reply by camelman90** [reply]
You can select the top of the hierarchy in Spoke and then "Override audio settings" to change user audio rolloff:
![image](https://user-images.githubusercontent.com/28685654/95786527-0ae23c00-0cd8-11eb-9f85-b40db726371a.png)

**Comment by jcklpe** [comment]
Also if there are any ways I can contribute, please let me know. I can do css/html stuff easily. I'm okay with react and javascript stuff though hardly the best. I know enough though to participate though. I help build design systems which are made in react, for a living. And like I said I also have 3D art experience. 

Took a stab at Spoke the other day and it seems serviceable though I think I would probably end up doing almost all my level design and editing in blender directly for the sake of performance reasons.

**Comment by pattersonbl2** [comment, marked as answer]
Opened a new issue for the Webcam view: https://github.com/mozilla/hubs/issues/4489 

We just released a new feature called audio zones , where you can change the audio fall off.  

https://hubs.mozilla.com/docs/spoke-adding-scene-content.html#audio-zones

https://github.com/MozillaReality/hubs-blender-exporter/releases/tag/v0_0_11
